okio/jvm/src/test/java/okio/PipeKotlinTest.kt

```diff
@@ -32,4 +49,682 @@ class PipeKotlinTest {
+      TimeUnit.MILLISECONDS.toNanos(100).toDouble())
+    assertEquals(underlyingOriginalDeadline, underlying.timeout().deadlineNanoTime())
+  }
+
+  @Test
+  fun honorsPipeSinkDeadlineOnClosingWhenUnderlyingSinkHasNoDeadline() {
+    val deadlineNanos = TimeUnit.MILLISECONDS.toNanos(500L)
+
+    val pipe = Pipe(4)
+    val underlying = TimeoutClosingSink()
+
+    underlying.timeout.clearDeadline()
+    pipe.sink.timeout().deadlineNanoTime(System.nanoTime() + deadlineNanos)
+
+    pipe.fold(underlying)
+
+    val start = System.currentTimeMillis()
+    pipe.sink.close()
+    val elapsed = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - start)
+
+    assertEquals(deadlineNanos.toDouble(), elapsed.toDouble(),
+      TimeUnit.MILLISECONDS.toNanos(100).toDouble())
+    assertFalse(underlying.timeout().hasDeadline())
+  }
+
+  @Test
+  fun honorsUnderlyingSinkDeadlineOnClosingWhenItIsSmaller() {
+    val smallerDeadlineNanos = TimeUnit.MILLISECONDS.toNanos(500L)
+    val biggerDeadlineNanos = TimeUnit.MILLISECONDS.toNanos(1500L)
+
+    val pipe = Pipe(4)
+    val underlying = TimeoutClosingSink()
+
+    val underlyingOriginalDeadline = System.nanoTime() + smallerDeadlineNanos
+    underlying.timeout.deadlineNanoTime(underlyingOriginalDeadline)
+    pipe.sink.timeout().deadlineNanoTime(System.nanoTime() + biggerDeadlineNanos)
+
+    pipe.fold(underlying)
+
+    val start = System.currentTimeMillis()
+    pipe.sink.close()
+    val elapsed = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - start)
+
+    assertEquals(smallerDeadlineNanos.toDouble(), elapsed.toDouble(),
+      TimeUnit.MILLISECONDS.toNanos(100).toDouble())
+    assertEquals(underlyingOriginalDeadline, underlying.timeout().deadlineNanoTime())
+  }
+
+  @Test
+  fun honorsUnderlyingSinkDeadlineOnClosingWhenPipeSinkHasNoDeadline() {
+    val deadlineNanos = TimeUnit.MILLISECONDS.toNanos(500L)
+
+    val pipe = Pipe(4)
+    val underlying = TimeoutClosingSink()
+
+    val underlyingOriginalDeadline = System.nanoTime() + deadlineNanos
+    underlying.timeout().deadlineNanoTime(underlyingOriginalDeadline)
+    pipe.sink.timeout().clearDeadline()
+
+    pipe.fold(underlying)
+
+    val start = System.currentTimeMillis()
+    pipe.sink.close()
+    val elapsed = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis() - start)
+
+    assertEquals(deadlineNanos.toDouble(), elapsed.toDouble(),
+      TimeUnit.MILLISECONDS.toNanos(100).toDouble())
+    assertEquals(underlyingOriginalDeadline, underlying.timeout().deadlineNanoTime())
+  }
+
+  /** Writes on this sink never complete. They can only time out. */
+  class TimeoutWritingSink : Sink {
+    val timeout = object : AsyncTimeout() {
+      override fun timedOut() {
+        synchronized(this@TimeoutWritingSink) {
+          (this@TimeoutWritingSink as Object).notifyAll()
+        }
+      }
+    }
+
+    override fun write(source: Buffer, byteCount: Long) {
```

Comment:
nit: to be correct to the Sink contract this method needs to call source.skip(byteCount)